engine/src/test/scala/cromwell/engine/workflow/CallCachingBlacklistManagerSpec.scala

```diff
@@ -0,0 +1,126 @@
+    id = WorkflowId.randomId(),
+    sources = workflowSourcesYesGrouping
+  )
+
+  val workflowYesGrouping2 = workflowYesGrouping1.copy(
+    id = WorkflowId.randomId()
+  )
+
+  it should "be off by default" in {
+    val configString = ""
+    val manager = new CallCachingBlacklistManager(ConfigFactory.parseString(configString), logger = NoLogging)
+
+    manager.blacklistCacheFor(workflowNoGrouping) shouldBe None
+  }
+
+  it should "be on with default values if blacklisting is enabled" in {
+    val configString =
+      """
+        |call-caching {
+        |  blacklist-cache {
+        |    enabled: true
+        |  }
+        |}
+        |""".stripMargin
+    val manager = new CallCachingBlacklistManager(ConfigFactory.parseString(configString), logger = NoLogging)
+
+    val cache = manager.blacklistCacheFor(workflowNoGrouping)
+    val rootWorkflowCache = cache.get.asInstanceOf[RootWorkflowBlacklistCache]
+
+    rootWorkflowCache.hitCache.size() shouldBe 0
+    rootWorkflowCache.bucketCache.size() shouldBe 0
+  }
+
+  it should "use root workflow level caches if no workflow-option is specified for groupings in config" in {
+    val configString =
+      """
+        |call-caching {
+        |  blacklist-cache {
+        |    enabled: true
+        |  }
+        |}
+        |""".stripMargin
+    val manager = new CallCachingBlacklistManager(ConfigFactory.parseString(configString), logger = NoLogging)
+    val cache = manager.blacklistCacheFor(workflowYesGrouping1)
+    val _ = cache.get.asInstanceOf[RootWorkflowBlacklistCache]
+  }
+
+  it should "use root workflow level caches if no workflow-option is provided in workflow options" in {
+    val configString =
+      """
+        |call-caching {
+        |  blacklist-cache {
+        |    enabled: true
+        |    groupings: {
+        |      workflow-option: "google_project"
+        |    }
+        |  }
+        |}
+        |""".stripMargin
+    val manager = new CallCachingBlacklistManager(ConfigFactory.parseString(configString), logger = NoLogging)
+    val cache = manager.blacklistCacheFor(workflowNoGrouping)
+    val _ = cache.get.asInstanceOf[RootWorkflowBlacklistCache]
+  }
+
+  it should "use a grouping cache if there is a workflow-option in config and its value exists as a key in workflow options" in {
+    val configString =
+      """
+        |call-caching {
+        |  blacklist-cache {
+        |    enabled: true
+        |    groupings: {
+        |      workflow-option: "google_project"
+        |    }
+        |  }
+        |}
+        |""".stripMargin
+    val manager = new CallCachingBlacklistManager(ConfigFactory.parseString(configString), logger = NoLogging)
+    val cache1 = manager.blacklistCacheFor(workflowYesGrouping1).get
+    val _ = cache1.asInstanceOf[GroupingBlacklistCache]
+
+    val cache2 = manager.blacklistCacheFor(workflowYesGrouping2).get
+    System.identityHashCode(cache1) shouldEqual System.identityHashCode(cache2)
+  }
```

Comment:
I guess some other other test cases:

* The cache entries should expire eventually (like the bucket/hit caches)
* If overfilled, the earlier cache entries are ejected